### PR TITLE
Revisit the code around onCompleted() in bidirectional streaming of gRPC

### PR DESCRIPTION
**Pull request by brfrn169**
This is a refactoring thing for the code around onCompleted() in bidirectional streaming of gRPC.

Please take a look!

### CODE REVIEW

File: server/src/main/java/com/scalar/db/server/DistributedStorageService.java [hunk 1]
```diff
@@ -206,8 +207,10 @@ public void onError(Throwable t) {
```
Comment (feeblefakie) [inline server/src/main/java/com/scalar/db/server/DistributedStorageService.java:207]: (Sorry, it cannot comment on the code, but it's for the `onError` method)
Not really sure but is it OK not to call `responseObserver.onError(t)` when onError is called?

File: server/src/main/java/com/scalar/db/server/DistributedStorageService.java [hunk 2]
```diff
@@ -206,8 +207,10 @@ public void onError(Throwable t) {
 
     @Override
     public void onCompleted() {
-      responseObserver.onCompleted();
-      cleanUp();
+      if (!cleanedUp.get()) {
```
Comment (feeblefakie) [inline server/src/main/java/com/scalar/db/server/DistributedStorageService.java:210]: I'm not sure but, is there any case where `cleanedUp` is set to true but `onCompleted` is not called?
If there is such a case, `onCompleted` won't be called properly?

File: core/src/main/java/com/scalar/db/storage/rpc/GrpcScanOnBidirectionalStream.java [hunk 3]
```diff
@@ -141,15 +141,15 @@ private void throwIfErrorForNext(ResponseOrError responseOrError) throws Executi
...
 
-  public void closeScanner() throws IOException {
+  public void closeScanner() throws ExecutionException {
     try {
       if (!hasMoreResults.get()) {
         return;
       }
-      requestObserver.onCompleted();
       hasMoreResults.set(false);
+      requestObserver.onCompleted();
```
Comment (feeblefakie) [inline core/src/main/java/com/scalar/db/storage/rpc/GrpcScanOnBidirectionalStream.java:150]: Just a question.
When users call `closeScanner` explicitly, this `onCompleted` will be called twice?
Is it OK to call `onCompleted` multiple times? (If OK, then I think there is no problem)

**Review by feeblefakie — COMMENTED** [review]
Overall looking good. Left several clarifying questions.

File: server/src/main/java/com/scalar/db/server/DistributedStorageService.java (hunk 2 above)
Comment (brfrn169) [reply server/src/main/java/com/scalar/db/server/DistributedStorageService.java:210]: The server side proactively cleans up only when the scanner has no more results. So when a user calls `closeScanner` before the server cleans up, the case will happen.

**Review by brfrn169 — COMMENTED** [review]
@feeblefakie Thank you for taking a loot at this!

I left some replies. Please check them when you have time!

**Review by feeblefakie — COMMENTED** [review]
Left additional questions. Sorry, I'm not familiar with the streaming...

**Review by feeblefakie — APPROVED** [review]
LGTM! Thank you!

**Pull request merged**